newIDE/app/src/EventsSheet/ParameterFields/StringWithSelectorField.js

```diff
@@ -1,38 +1,143 @@
 // @flow
-import * as React from 'react';
-import GenericExpressionField from './GenericExpressionField';
+import React from 'react';
+import { Trans, t } from '@lingui/macro';
 import {
   type ParameterFieldProps,
   type ParameterFieldInterface,
   type FieldFocusFunction,
 } from './ParameterFieldCommons';
-import { getParameterChoices } from './ParameterMetadataTools';
+import SelectField, { type SelectFieldInterface } from '../../UI/SelectField';
+
+import GenericExpressionField from './GenericExpressionField';
+import SelectOption from '../../UI/SelectOption';
+import { TextFieldWithButtonLayout } from '../../UI/Layout';
+import RaisedButton from '../../UI/RaisedButton';
+import Functions from '@material-ui/icons/Functions';
+import FlatButton from '../../UI/FlatButton';
+import TypeCursorSelect from '../../UI/CustomSvgIcons/TypeCursorSelect';
+import { getParameterChoiceValues } from './ParameterMetadataTools';
 
 export default React.forwardRef<ParameterFieldProps, ParameterFieldInterface>(
   function StringWithSelectorField(props: ParameterFieldProps, ref) {
-    const field = React.useRef<?GenericExpressionField>(null);
+    const field = React.useRef<?(
+      | GenericExpressionField
+      | SelectFieldInterface
+    )>(null);
+
     const focus: FieldFocusFunction = options => {
       if (field.current) field.current.focus(options);
     };
     React.useImperativeHandle(ref, () => ({
       focus,
     }));
 
+    const choices = getParameterChoiceValues(props.parameterMetadata);
+
+    const isCurrentValueInList = choices.some(
+      choice => `"${choice}"` === props.value
+    );
+
+    // If the current value is not in the list, display an expression field.
+    const [isExpressionField, setIsExpressionField] = React.useState(
+      !!props.value && !isCurrentValueInList
+    );
+
+    React.useEffect(
+      () => {
+        if (!isExpressionField && !props.value && choices.length > 0) {
+          props.onChange(`"${choices[0]}"`);
+        }
+      },
+      [choices, isExpressionField, props]
+    );
```

Comment:
`props` is a new object at every render, so I think putting it in the array of an effect is an anti pattern. Instead, pass what is really used:

```suggestion
    React.useEffect(
      () => {
        if (!isExpressionField && !props.value && choices.length > 0) {
          props.onChange(`"${choices[0]}"`);
        }
      },
      [choices, isExpressionField, props.onChange, props.onValue]
    );
```

It might be worth useMemo on `choices` too to make it stable. isExpressionField is fine.

Reply:
Actually, could we avoid the effect entirely? Like, just do call onChange when a choice is made maybe?

Reply:
> It might be worth useMemo on choices too to make it stable. isExpressionField is fine.

Other fields don't use a memo. I guess it's because the list can be changed in another tab without this component to know. For instance, for a layer field, a layer could be added in the scene editor while a parameter field is still living in the event editor.